File: capstone_project/web/src/services/inferenceService.js
// web/src/services/inferenceService.js
// Loads the exported TensorFlow.js graph model and runs BiLSTM inference.
// Falls back gracefully (this.loaded === false) when the model file is absent —
// the app then runs in rule-based-only mode.

import * as tf from '@tensorflow/tfjs';
import '@tensorflow/tfjs-backend-webgl';
import { CONST } from '../utils/constants.js';

const MODEL_PATH = './assets/models/model.json';

export class InferenceService {
  constructor() {
    this._model = null;
    // Rolling 30-frame window: Array of 12-element feature arrays
    this._window = [];
  }

  /** True when a model has been loaded successfully. */
  get loaded() { return this._model !== null; }

  /**
   * Load the TFJS graph model.
   * Silently fails if the model file is not present.
   * @param {(progress: number) => void} onProgress  0–1
   */
  async load(onProgress) {
    try {
      await tf.setBackend('webgl');
      await tf.ready();
      onProgress?.(0.2);

      this._model = await tf.loadLayersModel(MODEL_PATH);
      onProgress?.(0.8);

      // Warm-up prediction to avoid first-inference latency
      const dummy = tf.zeros([1, CONST.WINDOW_FRAMES, 12]);
      const warmup = this._model.predict(dummy);
      warmup.dispose();
      dummy.dispose();

      onProgress?.(1);
      console.log(`[Inference] Model loaded. Backend: ${tf.getBackend()}`);
    } catch (err) {
      this._model = null;
      console.warn(
        '[Inference] Model not found — running in rule-based mode.\n' +
        'Train the model via the Colab notebook and place model.json in web/assets/models/',
        err.message
      );
    }
  }

  /**
   * Add a 12-dim feature frame to the rolling window and run inference
   * when the window is full (30 frames).
   *
   * @param {number[]} features  12-element array
   * @returns {{ label: string, confidence: number, probs: Object } | null}
   */
  run(features) {
    if (!this._model) return null;

    this._window.push(features);
    if (this._window.length > CONST.WINDOW_FRAMES) this._window.shift();
    if (this._window.length < CONST.WINDOW_FRAMES) return null;

    // Build tensor: shape [1, 30, 12]
    let result = null;
    const inputTensor = tf.tensor3d([this._window], [1, CONST.WINDOW_FRAMES, 12]);
    try {
      const outputTensor = this._model.predict(inputTensor);
      const probs = Array.from(outputTensor.dataSync());
      outputTensor.dispose();

      const maxIdx = probs.indexOf(Math.max(...probs));
      if (probs[maxIdx] >= CONST.CLASSIFIER_CONFIDENCE) {
        result = {
          label:      CONST.CLASS_LABELS[maxIdx],
          confidence: probs[maxIdx],
          probs:      Object.fromEntries(CONST.CLASS_LABELS.map((l, i) => [l, probs[i]])),
        };
      }
    } catch (err) {
      console.error('[Inference] Prediction error:', err);
    } finally {
      inputTensor.dispose();
    }
    return result;
  }

  /** Clear the rolling feature window (call at session start). */
  reset() { this._window = []; }
}

File: capstone_project/lib/services/inference_service.dart
import 'package:flutter/foundation.dart';
import 'package:tflite_flutter/tflite_flutter.dart';
import '../core/constants/app_constants.dart';

class InferenceService {
  Interpreter? _interpreter;
  bool _modelLoaded = false;

  bool get modelLoaded => _modelLoaded;

  // Rolling window: [windowSize][features]
  final List<List<double>> _window = [];

  Future<void> initialize() async {
    try {
      final options = InterpreterOptions()..threads = 2;
      _interpreter = await Interpreter.fromAsset(
        AppConstants.modelAssetPath,
        options: options,
      );
      _modelLoaded = true;
      debugPrint('[InferenceService] TFLite model loaded ✓');
    } catch (e) {
      _modelLoaded = false;
      debugPrint('[InferenceService] No model found — running rule-based mode. ($e)');
    }
  }

  /// Add a 12-dim feature frame to the rolling window.
  /// Returns a [ClassifierResult] when the window is full (30 frames),
  /// or null if the window is still filling or model not loaded.
  ClassifierResult? processFrame(List<double> features) {
    if (!_modelLoaded || _interpreter == null) return null;

    _window.add(features);
    if (_window.length > AppConstants.classifierWindowFrames) {
      _window.removeAt(0);
    }
    if (_window.length < AppConstants.classifierWindowFrames) return null;

    // Build input tensor: shape [1, 30, 12]
    final input = [_window.map((f) => f).toList()];
    final output = List.generate(1, (_) => List.filled(AppConstants.classLabels.length, 0.0));

    try {
      _interpreter!.run(input, output);
    } catch (e) {
      debugPrint('[InferenceService] Inference error: $e');
      return null;
    }

    final probs = output[0];
    int maxIdx = 0;
    for (var i = 1; i < probs.length; i++) {
      if (probs[i] > probs[maxIdx]) maxIdx = i;
    }

    if (probs[maxIdx] < AppConstants.classifierConfidenceThreshold) return null;

    return ClassifierResult(
      label: AppConstants.classLabels[maxIdx],
      confidence: probs[maxIdx],
      allProbabilities: Map.fromIterables(AppConstants.classLabels, probs),
    );
  }

  void reset() => _window.clear();

  void dispose() {
    _interpreter?.close();
    _interpreter = null;
    _modelLoaded = false;
  }
}

class ClassifierResult {
  final String label;
  final double confidence;
  final Map<String, double> allProbabilities;
  const ClassifierResult({
    required this.label,
    required this.confidence,
    required this.allProbabilities,
  });
}
